File: java/formula-java/src/test/java/org/formula/test/FormulaTestCase.java
package org.formula.test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.fail;

import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.List;
import org.formula.ResolvedValue;
import org.formula.context.DataContext;

public record FormulaTestCase(String name,
                              String formula,
                              DataContext data,
                              @JsonProperty("expected_text") String expectedText,
                              @JsonProperty("expected_number") Number expectedNumber,
                              @JsonProperty("expected_boolean") Boolean expectedBoolean,
                              @JsonProperty("expected_error") String expectedError,
                              @JsonProperty("expected_name") String expectedName,
                              @JsonProperty("expected_list") List<ResolvedValue> expectedList) {

    public void assertResult(ResolvedValue result) {
        if (expectedError != null) {
            fail("Expected error, but none occurred");
        }
        if (expectedNumber != null && expectedNumber instanceof Integer i) {
            assertThat(result.asNumber()).isEqualTo(i);
        }
        if (expectedNumber != null && expectedNumber instanceof Double d) {
            assertThat(result.asDecimal()).isEqualTo(d);
        }
        if (expectedNumber != null && expectedNumber instanceof Float f) {
            assertThat(result.asDecimal()).isEqualTo(f);
        }
        if (expectedNumber != null && expectedNumber instanceof Long l) {
            assertThat(result.asNumber()).isEqualTo(l);
        }
        if (expectedText != null) {
            assertThat(result.asText()).isEqualTo(expectedText);
        }
        if (expectedBoolean != null) {
            assertThat(result.asBoolean()).isEqualTo(expectedBoolean);
        }
        if (expectedList != null) {
            assertThat(result.asList()).isEqualTo(expectedList);
        }
    }
}

File: formula-py/tests/formula_test_case.py
import math

import formula
from formula import resolved_value
from formula.resolvable import ResolvableList
from formula.static_resolvable import StaticResolvable


def _parse_value(value: str):
    if isinstance(value, str) and value.startswith('{') and value.endswith('}'):
        return formula.parse(value[1:-1])
    if isinstance(value, list):
        return list(map(resolved_value, value))
    return StaticResolvable(value)


def _parse_data(data: dict):
    if data is None:
        return {}
    for key in data.keys():
        value = data.get(key)
        if isinstance(value, str) and value.startswith('{') and value.endswith('}'):
            data[key] = formula.parse(value[1:-1])
        elif isinstance(value, list):
            data[key] = ResolvableList(map(_parse_value, value))
        elif value is not None:
            data[key] = StaticResolvable(value)
    return data


class FormulaTestCase:
    def __init__(self, yaml: dict):
        self.name = yaml.get('name')
        self.formula = yaml.get('formula')
        self.data = _parse_data(yaml.get('data'))
        self.expected_text = yaml.get('expected_text')
        self.expected_number = yaml.get('expected_number')
        if self.expected_number == 'NaN':
            self.expected_number = math.nan
        self.expected_boolean = yaml.get('expected_boolean')
        self.expected_error = yaml.get('expected_error')
        self.expected_list = yaml.get('expected_list')

    def __repr__(self):
        return self.name
